### ISSUE TITLE
Some way to specify expected failures for OpInfo-based tests

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#62354]
## 🚀 Feature

A way to specify expected failure for OpInfo-based tests that works with unittest's expected failure reporting

## Motivation

For functorch we have automated coverage tests for OpInfos. E.g. there is a test that for each OpInfo, checks vmap support for the OpInfo. Tests that are currently failing are put into a denylist and skipped instead of running. It would be nice if we could mark them as expected failure instead.

## Alternatives

We can emulate expected failures by doing a try-except and checking for the presence of an exception. The main problem I have with this approach is the UX: the unittest runner shows an "expected failure" as a passed test, when it would be nice to see that it is an expected failure in the runner.

cc @mruberry @VitalyFedyunin @walterddr

**Comment by mruberry** [comment]
cc @heitorschueroff, who's adding an efail mechanism, although I'm not sure how well that mechanism will work for out-of-core projects

**Comment by heitorschueroff** [comment]
@zou3519 I have this PR https://github.com/pytorch/pytorch/pull/62198 though as @mruberry suggested, I will change the SkipInfo instead to default to expected failure so that all existing skips can be tested for whether they are still failing.

**Comment by mruberry** [comment]
I suppose one option would be to acquire the OpInfos that you want to mark as failing the test(s) and add the desired SkipInfo to them?

**Comment by heitorschueroff** [comment]
support for expected failure added here https://github.com/pytorch/pytorch/pull/62963

**Comment by zou3519** [comment]
Reopening this for a related feature request. The existing expected failure mechanism is great, but we cannot use it in functorch because it is out-of-tree. We would like to be able to modify the SkipInfo of an operator to include a test that we've defined out of tree. For example, [here](https://github.com/facebookresearch/functorch/blob/cdba85ddd7159640211924f0b512625d945c0f0e/test/test_ops.py#L318-L381) we'd like to change this large list of skips into expected failures.

Would it be possible to have a pytorch-blessed mechanism of doing this? I can manually grab OpInfo and edit its decorators, but I'm worried that this will break without being officially supported.

**Comment by heitorschueroff** [comment]
@zou3519 The decorators get applied here https://github.com/pytorch/pytorch/blob/4a390a56c4a345a917f5a4d8ee6cd47818fa7e84/torch/testing/_internal/common_device_type.py#L768

I see the following options:

1. You could add a SkipInfo(...) to the operators in op_db of common_methods_invocations.py to target your test, in this case, `SkipInfo('TestOperators', 'test_vmapvjp')`. The decorator only gets applied if the class and test name match, otherwise it is ignored. This will work fine as long as no other test in-tree matches the same class and test name.
2. You could, as you said, copy the opinfos from op_db and manually edit their decorators field.
3. We could add a kwarg to @ops that accepts a function to apply custom decorators to the operators. This way, you could provide a function that applies the `unittest.expectedFailure` decorator to the operators in the list of skips.

If you think option 3 is the way to go I can add support for it.

PS: I'm working on changing SkipInfo to an XFAIL if expected_failure=True.

cc @mruberry

**Comment by zou3519** [comment]
> You could add a SkipInfo(...) to the operators in op_db of common_methods_invocations.py to target your test, in this case, SkipInfo('TestOperators', 'test_vmapvjp'). The decorator only gets applied if the class and test name match, otherwise it is ignored. This will work fine as long as no other test in-tree matches the same class and test name.

The downside to doing this is that in order to skip a test in functorch, we have to land a PR to pytorch/pytorch.

> You could, as you said, copy the opinfos from op_db and manually edit their decorators field.

I think the question here is, how common do you think out-of-tree projects will use OpInfos? If the answer is "more than just functorch", then it's worth having an API to do this because it is a potentially useful operation to be able to skip tests and mark them as expected failure.

> We could add a kwarg to @ops that accepts a function to apply custom decorators to the operators. This way, you could provide a function that applies the unittest.expectedFailure decorator to the operators in the list of skips.

I like this option the best. There are a couple of other things we could do in this design space. Perhaps we could add an @opsSkip decorator that lets one directly add a skip:
```
@ops(ops_db)
@opsCustomSkip([
    ('sin',),  # op_name
    ('moveaxis', 'variant_name'),  # op_name, variant_name
])
def test_vmapvjp(self, ...)
```
The custom decorator idea might be better because there's more flexibility

**Comment by heitorschueroff** [comment]
> I like this option the best. There are a couple of other things we could do in this design space. Perhaps we could add an @opsSkip decorator that lets one directly add a skip:
> 
> ```
> @ops(ops_db)
> @opsCustomSkip([
>     ('sin',),  # op_name
>     ('moveaxis', 'variant_name'),  # op_name, variant_name
> ])
> def test_vmapvjp(self, ...)
> ```
> 
> The custom decorator idea might be better because there's more flexibility

I agree with you that the third option is the best. However, I'm not entirely sure we can have multiple decorators as of now. One way it might work is to have `@opsSkip` custom decorator store its arguments into the function being decorated and then have `@ops` read the stored attributes in the `__call__` method. But I'm not seeing how this is more flexible then passing a function to `@ops` that accepts test, op, device and dtype and returns a decorated test.

For reference, @jbschlosser has been working on a generic parametrization mechanism here https://github.com/pytorch/pytorch/pull/60753.

**Comment by mruberry** [comment]
Our recommendation for third parties using OpInfos is generally to implement skipping by looking up the opinfo when the test starts and skipping it then or by prefiltering with an allow list like like PyTorch/XLA does:

https://github.com/pytorch/xla/blob/59739f2e298dd40e8a33cecc08d4ec42c91f66fe/test/test_ops.py#L14

Their opt-in approach means their test doesn't break when new OpInfos are added, and they don't need to change PyTorch, either. 

In the future we could look at extending torch.testing to allow acquisition of OpInfos or properties of them (like what dtypes an op supports), but editing them would be tricky to support. It would be nice if we could raise an efail signal from a test like we can raise skips, too, but I can't think of a good way to do that at the moment. 

So adding decorators is makes sense. What we could do is extend `@ops` with a kwarg like `dynamic_decorators` that would point to a function. The signature of this function could be:

```
dynamic_decorators(opinfo, device, dtype, **kwargs) -> Sequence of Decorators
```

and it would have to return a sequence of decorators that could wrap the test. So then this decorator could be called when the test template is instantiated and its additional decorators applied to the test. 

We'd have to very carefully document the extensibility point but the actual implementation is straightforward, I think.
